### PR TITLE
Warn about dead tuple struct fields

### LINKED ISSUE #92790: dead_code lint does not trigger for tuple structs with unread fields
This code:

```rust
fn main() {
    let _ = Example { u: 0 };
}

struct Example {
    u: u8,
}
```

Produces this warning:

```
warning: field is never read: `u`
 --> src/main.rs:6:5
  |
6 |     u: u8,
  |     ^^^^^
  |
```

However, switching to a tuple struct:

```rust
fn main() {
    let _ = Example(0);
}

struct Example(u8);
```

Does not report a warning.

<details>
<summary>Original report misleadingly focusing on <code>Debug</code></summary>

```rust
fn main() {
    dbg!(Example { a: 0 });
}

#[derive(Debug)]
struct Example {
    a: u8,
}
```

produces 

```
warning: field is never read: `a`
 --> src/main.rs:7:5
  |
7 |     a: u8,
  |     ^^^^^
  |
  = note: `#[warn(dead_code)]` on by default
```

However, switching to a tuple struct:

```rust
fn main() {
    dbg!(Example(0));
}

#[derive(Debug)]
struct Example(u8);
```

does not produce the warning.

/cc @FabianWolff
/cc #85200; #84647; #88900

</details>

### Meta
Rustc 1.57.0 and 1.60.0-nightly (2022-01-10 89b9f7b284aacc5f8613)

### PR DESCRIPTION
**Pull request by FabianWolff**
Fixes #92790. Since the introduction of warnings about unused fields (commit https://github.com/rust-lang/rust/commit/0271224bdae26260ab498f47323997f9edb5879e in 2014), only _named_ fields got warnings, but not positional fields in tuple structs or tuple enum variants.

Warning about positional fields is probably somewhat controversial because
1. It is harder to fix than removing named fields, because removing a positional field changes the indices of the following fields, and pattern matches will also have to be adapted.
2. It will uncover lots of dead fields in existing code, as it already did in the compiler code, which will lead to similar issues as in #85200/#88900.

I am putting this up for discussion now. Maybe this should be nominated for language team sign-off as well?

cc @shepmaster
r? @camelid

### CODE REVIEW

File: src/librustdoc/clean/types.rs [hunk 1]
```diff
@@ -2312,7 +2312,7 @@ impl TypeBinding {
 crate enum SubstParam {
     Type(Type),
     Lifetime(Lifetime),
-    Constant(Constant),
+    Constant,
```
Comment (camelid) [inline src/librustdoc/clean/types.rs:2315]: Hmm... I'm not sure about this change. I'll have to look into it.

File: compiler/rustc_apfloat/src/ppc.rs [hunk 2]
```diff
@@ -30,7 +30,7 @@ pub type DoubleDouble = DoubleFloat<ieee::Double>;
 // FIXME: Implement all operations in DoubleDouble, and delete these
 // semantics.
 // FIXME(eddyb) This shouldn't need to be `pub`, it's only used in bounds.
-pub struct FallbackS<F>(F);
+pub struct FallbackS<F>(#[allow(dead_code)] F);
```
Comment (eddyb) [inline compiler/rustc_apfloat/src/ppc.rs:33]: I wish we had a better name for this than "dead code". I'm actually surprised it's not "unused" instead.

I suppose this change is small enough, and just a lint attribute, that it's probably fine.

Comment (FabianWolff) [reply compiler/rustc_apfloat/src/ppc.rs:33]: Well, you could use `unused` as well, but that is the lint group containing the `dead_code` lint, so the latter is more precise (not that it makes a difference here). Also, just to be clear: I didn't come up with the name here, it is also called `dead_code` for named fields.

In any case: Thanks for having a look!

**Comment by camelid** [comment]
I don't think I have enough knowledge of this code nor time to review this PR in its entirety, but I'll try to review some of rustdoc changes when I get a chance. Thanks for working on this! :)

r? rust-lang/compiler

**Comment by joshtriplett** [comment]
I do think this needs a lang signoff; nominating.

**Comment by joshtriplett** [comment]
Given that this may be harder for people to change, I think it deserves a separate sub-lint of `dead_code`, such as `dead_tuple_field`.

I do think this is a good idea, and we should make this change.

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/lang meeting. We'd like to go with a gradual approach:

1) Introduce this as `dead_tuple_field`, allow-by-default at first.
2) Announce this, and tell people to try it.
3) Later, bump it up to warn. People will have the option of disabling `dead_tuple_field` specifically, if they need to.

Thanks for working on this!

**Comment by pnkfelix** [comment]
I couldn't tell from skimming the PR: What does this warning do for fields of unit type, `()`?

I ask because my first reaction when I read:
> 
> Warning about positional fields is probably somewhat controversial because
> 1. It is harder to fix than removing named fields, because removing a positional field changes the indices of the following fields, and pattern matches will also have to be adapted.
> [...]

is that I thought: "well, we should let people include `()` fields (i.e. not warn about them), and perhaps even have the lint diagnostic here suggest either removing the field or making it have `()` type. That keep the indices unchanged."

But, like I said above, I cannot tell from a quick read what this does for `()`. (I did see it does something special for unit structs (`struct S;`), but that is a slightly different situation and the handling there is to address a different problem.)

**Comment by JohnCSimon** [comment]
Ping from triage: 
@FabianWolff can you please address the comments, merge conflicts, build issues, and post your status on this PR?

**Comment by Dylan-DPC** [comment]
closing this as inactive

**Comment by FabianWolff** [comment]
Ah, I didn't know this, I thought I was lacking the necessary permissions for reopening. Thanks for the hint @mati865!

**Comment by shepmaster** [comment]
@Dylan-DPC  you should mention the force-push caveat as part of your canned closing message.

**Pull request closed without merging**